// @clickhouse/client-common/dist/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClickHouseLogLevel = exports.LogWriter = exports.DefaultLogger = void 0;
class DefaultLogger {
    trace({ module, message, args }) {
        const params = [
            formatMessage({ module, message, level: 'TRACE' }),
        ];
        if (args) {
            params.push('\nArguments:', args);
        }
        console.debug(...params);
    }
    debug({ module, message, args }) {
        const params = [
            formatMessage({ module, message, level: 'DEBUG' }),
        ];
        if (args) {
            params.push('\nArguments:', args);
        }
        console.debug(...params);
    }
    info({ module, message, args }) {
        const params = [
            formatMessage({ module, message, level: 'INFO' }),
        ];
        if (args) {
            params.push('\nArguments:', args);
        }
        console.info(...params);
    }
    warn({ module, message, args, err }) {
        const params = [
            formatMessage({ module, message, level: 'WARN' }),
        ];
        if (args) {
            params.push('\nArguments:', args);
        }
        if (err) {
            params.push('\nCaused by:', err);
        }
        console.warn(...params);
    }
    error({ module, message, args, err }) {
        const params = [
            formatMessage({ module, message, level: 'ERROR' }),
        ];
        if (args) {
            params.push('\nArguments:', args);
        }
        params.push('\nCaused by:', err);
        console.error(...params);
    }
}
exports.DefaultLogger = DefaultLogger;
class LogWriter {
    logger;
    module;
    logLevel;
    constructor(logger, module, logLevel) {
        this.logger = logger;
        this.module = module;
        this.logLevel = logLevel;
        this.info({
            message: `Log level is set to ${ClickHouseLogLevel[this.logLevel]}`,
        });
    }
    trace(params) {
        if (this.logLevel <= ClickHouseLogLevel.TRACE) {
            this.logger.trace({
                ...params,
                module: params.module ?? this.module,
            });
        }
    }
    debug(params) {
        if (this.logLevel <= ClickHouseLogLevel.DEBUG) {
            this.logger.debug({
                ...params,
                module: params.module ?? this.module,
            });
        }
    }
    info(params) {
        if (this.logLevel <= ClickHouseLogLevel.INFO) {
            this.logger.info({
                ...params,
                module: params.module ?? this.module,
            });
        }
    }
    warn(params) {
        if (this.logLevel <= ClickHouseLogLevel.WARN) {
            this.logger.warn({
                ...params,
                module: params.module ?? this.module,
            });
        }
    }
    error(params) {
        if (this.logLevel <= ClickHouseLogLevel.ERROR) {
            this.logger.error({
                ...params,
                module: params.module ?? this.module,
            });
        }
    }
}
exports.LogWriter = LogWriter;
var ClickHouseLogLevel;
(function (ClickHouseLogLevel) {
    /**
     * A fine-grained debugging event. Might produce a lot of logs, so use with caution.
     */
    ClickHouseLogLevel[ClickHouseLogLevel["TRACE"] = 0] = "TRACE";
    /**
     * A debugging event. Useful for debugging, but generally not needed in production. Includes technical values that might require redacting.
     */
    ClickHouseLogLevel[ClickHouseLogLevel["DEBUG"] = 1] = "DEBUG";
    /**
     * An informational event. Indicates that an event happened.
     */
    ClickHouseLogLevel[ClickHouseLogLevel["INFO"] = 2] = "INFO";
    /**
     * A warning event. Not an error, but is likely more important than an informational event. Addressing should help prevent potential issues.
     */
    ClickHouseLogLevel[ClickHouseLogLevel["WARN"] = 3] = "WARN";
    /**
     * An error event. Something went wrong.
     */
    ClickHouseLogLevel[ClickHouseLogLevel["ERROR"] = 4] = "ERROR";
    /**
     * Logging is turned off.
     */
    ClickHouseLogLevel[ClickHouseLogLevel["OFF"] = 127] = "OFF";
})(ClickHouseLogLevel || (exports.ClickHouseLogLevel = ClickHouseLogLevel = {}));
function formatMessage({ level, module, message, }) {
    const ts = new Date().toISOString();
    return `[${ts}][${level}][@clickhouse/client][${module}] ${message}`;
}
