// @clickhouse/client/dist/connection/node_base_connection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeBaseConnection = void 0;
const client_common_1 = require("@clickhouse/client-common");
const crypto_1 = __importDefault(require("crypto"));
const stream_1 = __importDefault(require("stream"));
const zlib_1 = __importDefault(require("zlib"));
const utils_1 = require("../utils");
const compression_1 = require("./compression");
const stream_2 = require("./stream");
class NodeBaseConnection {
    params;
    agent;
    defaultAuthHeader;
    defaultHeaders;
    jsonHandling;
    knownSockets = new WeakMap();
    connectionId = crypto_1.default.randomUUID();
    socketCounter = 0;
    // For overflow concerns:
    //   node -e 'console.log(Number.MAX_SAFE_INTEGER / (1_000_000 * 60 * 60 * 24 * 366))'
    // gives 284 years of continuous operation at 1M requests per second
    // before overflowing the 53-bit integer
    requestCounter = 0;
    getNewSocketId() {
        this.socketCounter += 1;
        return `${this.connectionId}:${this.socketCounter}`;
    }
    getNewRequestId() {
        this.requestCounter += 1;
        return `${this.connectionId}:${this.requestCounter}`;
    }
    constructor(params, agent) {
        this.params = params;
        this.agent = agent;
        if (params.auth.type === 'Credentials') {
            this.defaultAuthHeader = `Basic ${Buffer.from(`${params.auth.username}:${params.auth.password}`).toString('base64')}`;
        }
        else if (params.auth.type === 'JWT') {
            this.defaultAuthHeader = `Bearer ${params.auth.access_token}`;
        }
        else {
            throw new Error(`Unknown auth type: ${params.auth.type}`);
        }
        this.defaultHeaders = {
            // Node.js HTTP agent, for some reason, does not set this on its own when KeepAlive is enabled
            Connection: this.params.keep_alive.enabled ? 'keep-alive' : 'close',
            'User-Agent': (0, utils_1.getUserAgent)(this.params.application_id),
        };
        this.jsonHandling = params.json ?? {
            parse: JSON.parse,
            stringify: JSON.stringify,
        };
    }
    async ping(params) {
        const { log_writer, log_level } = this.params;
        const query_id = this.getQueryId(params.query_id);
        const { controller, controllerCleanup } = this.getAbortController(params);
        try {
            let result;
            if (params.select) {
                const searchParams = (0, client_common_1.toSearchParams)({
                    database: undefined,
                    query: PingQuery,
                    query_id,
                });
                result = await this.request({
                    query: PingQuery,
                    method: 'GET',
                    url: (0, client_common_1.transformUrl)({ url: this.params.url, searchParams }),
                    abort_signal: controller.signal,
                    headers: this.buildRequestHeaders(),
                    query_id,
                    log_writer,
                    log_level,
                }, 'Ping');
            }
            else {
                result = await this.request({
                    query: 'ping',
                    method: 'GET',
                    url: (0, client_common_1.transformUrl)({ url: this.params.url, pathname: '/ping' }),
                    abort_signal: controller.signal,
                    headers: this.buildRequestHeaders(),
                    query_id,
                    log_writer,
                    log_level,
                }, 'Ping');
            }
            await (0, stream_2.drainStreamInternal)({
                op: 'Ping',
                log_writer,
                query_id,
                log_level,
            }, result.stream);
            return { success: true };
        }
        catch (error) {
            // it is used to ensure that the outgoing request is terminated,
            // and we don't get unhandled error propagation later
            controller.abort('Ping failed');
            // not an error, as this might be semi-expected
            if (log_level <= client_common_1.ClickHouseLogLevel.WARN) {
                log_writer.warn({
                    message: this.httpRequestErrorMessage('Ping'),
                    err: error,
                    args: {
                        connection_id: this.connectionId,
                        query_id,
                    },
                });
            }
            return {
                success: false,
                error: error, // should NOT be propagated to the user
            };
        }
        finally {
            controllerCleanup();
        }
    }
    async query(params) {
        const { log_writer, log_level } = this.params;
        const query_id = this.getQueryId(params.query_id);
        const clickhouse_settings = (0, client_common_1.withHttpSettings)(params.clickhouse_settings, this.params.compression.decompress_response);
        const searchParams = (0, client_common_1.toSearchParams)({
            database: this.params.database,
            query_params: params.query_params,
            session_id: params.session_id,
            clickhouse_settings,
            query_id,
            role: params.role,
        });
        const { controller, controllerCleanup } = this.getAbortController(params);
        // allows enforcing the compression via the settings even if the client instance has it disabled
        const enableResponseCompression = clickhouse_settings.enable_http_compression === 1;
        try {
            const { response_headers, stream, http_status_code } = await this.request({
                method: 'POST',
                url: (0, client_common_1.transformUrl)({ url: this.params.url, searchParams }),
                body: params.query,
                abort_signal: controller.signal,
                enable_response_compression: enableResponseCompression,
                headers: this.buildRequestHeaders(params),
                query: params.query,
                query_id,
                log_writer,
                log_level,
            }, 'Query');
            return {
                stream,
                response_headers,
                query_id,
                http_status_code,
            };
        }
        catch (err) {
            controller.abort('Query HTTP request failed');
            this.logRequestError({
                op: 'Query',
                query_id: query_id,
                query_params: params,
                search_params: searchParams,
                err: err,
                extra_args: {
                    decompress_response: enableResponseCompression,
                    clickhouse_settings,
                },
            });
            throw err; // should be propagated to the user
        }
        finally {
            controllerCleanup();
        }
    }
    async insert(params) {
        const { log_writer, log_level } = this.params;
        const query_id = this.getQueryId(params.query_id);
        const searchParams = (0, client_common_1.toSearchParams)({
            database: this.params.database,
            clickhouse_settings: params.clickhouse_settings,
            query_params: params.query_params,
            query: params.query,
            session_id: params.session_id,
            role: params.role,
            query_id,
        });
        const { controller, controllerCleanup } = this.getAbortController(params);
        try {
            const { stream, summary, response_headers, http_status_code } = await this.request({
                method: 'POST',
                url: (0, client_common_1.transformUrl)({ url: this.params.url, searchParams }),
                body: params.values,
                abort_signal: controller.signal,
                enable_request_compression: this.params.compression.compress_request,
                parse_summary: true,
                headers: this.buildRequestHeaders(params),
                query: params.query,
                query_id,
                log_writer,
                log_level,
            }, 'Insert');
            await (0, stream_2.drainStreamInternal)({
                op: 'Insert',
                log_writer,
                query_id,
                log_level,
            }, stream);
            return { query_id, summary, response_headers, http_status_code };
        }
        catch (err) {
            controller.abort('Insert HTTP request failed');
            this.logRequestError({
                op: 'Insert',
                query_id: query_id,
                query_params: params,
                search_params: searchParams,
                err: err,
                extra_args: {
                    clickhouse_settings: params.clickhouse_settings ?? {},
                },
            });
            throw err; // should be propagated to the user
        }
        finally {
            controllerCleanup();
        }
    }
    async exec(params) {
        const query_id = this.getQueryId(params.query_id);
        return this.runExec({
            ...params,
            query_id,
            op: 'Exec',
        });
    }
    async command(params) {
        const { log_writer, log_level } = this.params;
        const query_id = this.getQueryId(params.query_id);
        const commandStartTime = Date.now();
        if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
            log_writer.trace({
                message: 'Command: operation started',
                args: {
                    operation: 'Command',
                    connection_id: this.connectionId,
                    query_id,
                },
            });
        }
        const { stream, summary, response_headers } = await this.runExec({
            ...params,
            query_id,
            op: 'Command',
        });
        const runExecDuration = Date.now() - commandStartTime;
        if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
            log_writer.trace({
                message: 'Command: runExec completed, starting stream drain',
                args: {
                    operation: 'Command',
                    connection_id: this.connectionId,
                    query_id,
                    runExec_duration_ms: runExecDuration,
                    stream_state: {
                        readable: stream.readable,
                        readableEnded: stream.readableEnded,
                        readableLength: stream.readableLength,
                    },
                },
            });
        }
        // ignore the response stream and release the socket immediately
        const drainStartTime = Date.now();
        await (0, stream_2.drainStreamInternal)({
            op: 'Command',
            log_writer,
            query_id,
            log_level,
        }, stream);
        if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
            const drainDuration = Date.now() - drainStartTime;
            const totalDuration = Date.now() - commandStartTime;
            log_writer.trace({
                message: 'Command: operation completed',
                args: {
                    operation: 'Command',
                    connection_id: this.connectionId,
                    query_id,
                    drain_duration_ms: drainDuration,
                    total_duration_ms: totalDuration,
                },
            });
        }
        return { query_id, summary, response_headers };
    }
    async close() {
        if (this.agent !== undefined && this.agent.destroy !== undefined) {
            this.agent.destroy();
        }
    }
    defaultHeadersWithOverride(params) {
        return {
            // Custom HTTP headers from the client configuration
            ...(this.params.http_headers ?? {}),
            // Custom HTTP headers for this particular request; it will override the client configuration with the same keys
            ...(params?.http_headers ?? {}),
            // Includes the `Connection` + `User-Agent` headers which we do not allow to override
            // An appropriate `Authorization` header might be added later
            // It is not always required - see the TLS headers in `node_https_connection.ts`
            ...this.defaultHeaders,
        };
    }
    buildRequestHeaders(params) {
        const headers = this.defaultHeadersWithOverride(params);
        if ((0, client_common_1.isJWTAuth)(params?.auth)) {
            return {
                ...headers,
                Authorization: `Bearer ${params.auth.access_token}`,
            };
        }
        if (this.params.set_basic_auth_header) {
            if ((0, client_common_1.isCredentialsAuth)(params?.auth)) {
                return {
                    ...headers,
                    Authorization: `Basic ${Buffer.from(`${params.auth.username}:${params.auth.password}`).toString('base64')}`,
                };
            }
            else {
                return {
                    ...headers,
                    Authorization: this.defaultAuthHeader,
                };
            }
        }
        return {
            ...headers,
        };
    }
    getQueryId(query_id) {
        return query_id || crypto_1.default.randomUUID();
    }
    // a wrapper over the user's Signal to terminate the failed requests
    getAbortController(params) {
        const controller = new AbortController();
        function onAbort() {
            controller.abort();
        }
        params.abort_signal?.addEventListener('abort', onAbort);
        return {
            controller,
            controllerCleanup: () => {
                params.abort_signal?.removeEventListener('abort', onAbort);
            },
        };
    }
    logRequestError({ op, err, query_id, query_params, extra_args, }) {
        if (this.params.log_level <= client_common_1.ClickHouseLogLevel.ERROR) {
            this.params.log_writer.error({
                message: this.httpRequestErrorMessage(op),
                err: err,
                args: {
                    operation: op,
                    connection_id: this.connectionId,
                    query_id,
                    with_abort_signal: query_params.abort_signal !== undefined,
                    session_id: query_params.session_id,
                    ...extra_args,
                },
            });
        }
    }
    httpRequestErrorMessage(op) {
        return `${op}: HTTP request error.`;
    }
    parseSummary(op, response) {
        const summaryHeader = response.headers['x-clickhouse-summary'];
        if (typeof summaryHeader === 'string') {
            try {
                return this.jsonHandling.parse(summaryHeader);
            }
            catch (err) {
                if (this.params.log_level <= client_common_1.ClickHouseLogLevel.ERROR) {
                    this.params.log_writer.error({
                        message: `${op}: failed to parse X-ClickHouse-Summary header.`,
                        args: {
                            operation: op,
                            connection_id: this.connectionId,
                            'X-ClickHouse-Summary': summaryHeader,
                        },
                        err: err,
                    });
                }
            }
        }
    }
    async runExec(params) {
        const { log_writer, log_level } = this.params;
        const query_id = params.query_id;
        const sendQueryInParams = params.values !== undefined;
        const clickhouse_settings = (0, client_common_1.withHttpSettings)(params.clickhouse_settings, this.params.compression.decompress_response);
        const toSearchParamsOptions = {
            query: sendQueryInParams ? params.query : undefined,
            database: this.params.database,
            query_params: params.query_params,
            session_id: params.session_id,
            role: params.role,
            clickhouse_settings,
            query_id,
        };
        const searchParams = (0, client_common_1.toSearchParams)(toSearchParamsOptions);
        const { controller, controllerCleanup } = this.getAbortController(params);
        const tryDecompressResponseStream = params.op === 'Exec'
            ? // allows disabling stream decompression for the `Exec` operation only
                (params.decompress_response_stream ??
                    this.params.compression.decompress_response)
            : // there is nothing useful in the response stream for the `Command` operation,
                // and it is immediately destroyed; never decompress it
                false;
        const ignoreErrorResponse = params.ignore_error_response ?? false;
        try {
            const { stream, summary, response_headers, http_status_code } = await this.request({
                method: 'POST',
                url: (0, client_common_1.transformUrl)({ url: this.params.url, searchParams }),
                body: sendQueryInParams ? params.values : params.query,
                abort_signal: controller.signal,
                parse_summary: true,
                enable_request_compression: this.params.compression.compress_request,
                enable_response_compression: this.params.compression.decompress_response,
                try_decompress_response_stream: tryDecompressResponseStream,
                ignore_error_response: ignoreErrorResponse,
                headers: this.buildRequestHeaders(params),
                query: params.query,
                query_id,
                log_writer,
                log_level,
            }, params.op);
            return {
                stream,
                query_id,
                summary,
                response_headers,
                http_status_code,
            };
        }
        catch (err) {
            controller.abort(`${params.op} HTTP request failed`);
            this.logRequestError({
                op: params.op,
                query_id: query_id,
                query_params: params,
                search_params: searchParams,
                err: err,
                extra_args: {
                    clickhouse_settings: params.clickhouse_settings ?? {},
                },
            });
            throw err; // should be propagated to the user
        }
        finally {
            controllerCleanup();
        }
    }
    async request(params, op) {
        // allows the event loop to process the idle socket timers, if the CPU load is high
        // otherwise, we can occasionally get an expired socket, see https://github.com/ClickHouse/clickhouse-js/issues/294
        await (0, client_common_1.sleep)(0);
        const { log_writer, query_id, log_level } = params;
        const currentStackTrace = this.params.capture_enhanced_stack_trace
            ? (0, client_common_1.getCurrentStackTrace)()
            : undefined;
        const requestTimeout = this.params.request_timeout;
        return new Promise((resolve, reject) => {
            const start = Date.now();
            const request = this.createClientRequest(params);
            const request_id = this.getNewRequestId();
            const onError = (e) => {
                removeRequestListeners();
                if (e instanceof Error) {
                    if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
                        if (e.code === 'ECONNRESET') {
                            log_writer.trace({
                                message: `${op}: connection reset by peer`,
                                args: {
                                    operation: op,
                                    connection_id: this.connectionId,
                                    query_id,
                                    request_id,
                                },
                                module: 'HTTP Adapter',
                            });
                        }
                    }
                    if (log_level <= client_common_1.ClickHouseLogLevel.WARN) {
                        if (this.params.keep_alive.enabled) {
                            if (e.code === 'ECONNRESET') {
                                const socket = request.socket;
                                if (socket) {
                                    const socketInfo = this.knownSockets.get(socket);
                                    if (socketInfo) {
                                        const serverTimeoutMs = socketInfo.server_keep_alive_timeout_ms;
                                        if (serverTimeoutMs !== undefined) {
                                            if (this.params.keep_alive.idle_socket_ttl > serverTimeoutMs) {
                                                log_writer.warn({
                                                    message: `${op}: idle socket TTL is greater than server keep-alive timeout, try setting idle socket TTL to a value lower than the server keep-alive timeout to prevent unexpected connection resets, see https://c.house/js_keep_alive_econnreset for more details.`,
                                                    args: {
                                                        operation: op,
                                                        connection_id: this.connectionId,
                                                        query_id,
                                                        request_id,
                                                        socket_id: socketInfo.id,
                                                        server_keep_alive_timeout_ms: serverTimeoutMs,
                                                        idle_socket_ttl: this.params.keep_alive.idle_socket_ttl,
                                                    },
                                                    module: 'HTTP Adapter',
                                                });
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                    const err = (0, client_common_1.enhanceStackTrace)(e, currentStackTrace);
                    reject(err);
                }
                else {
                    reject(e);
                }
            };
            let responseStream;
            const onResponse = async (_response) => {
                if (this.params.log_level <= client_common_1.ClickHouseLogLevel.DEBUG) {
                    const duration = Date.now() - start;
                    this.params.log_writer.debug({
                        module: 'HTTP Adapter',
                        message: `${op}: got a response from ClickHouse`,
                        args: {
                            operation: op,
                            connection_id: this.connectionId,
                            query_id,
                            request_id,
                            request_method: params.method,
                            request_path: params.url.pathname,
                            response_status: _response.statusCode,
                            response_time_ms: duration,
                        },
                    });
                }
                if (this.params.keep_alive.enabled) {
                    const keepAliveHeader = _response.headers['keep-alive'];
                    if (keepAliveHeader) {
                        const [, timeout] = /timeout=(\d+)/i.exec(String(keepAliveHeader)) ?? [];
                        if (timeout) {
                            const socketInfo = this.knownSockets.get(_response.socket);
                            if (socketInfo) {
                                const timeoutMs = Number(timeout) * 1000;
                                socketInfo.server_keep_alive_timeout_ms = timeoutMs;
                                if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
                                    this.params.log_writer.trace({
                                        module: 'HTTP Adapter',
                                        message: `${op}: updated server sent socket keep-alive timeout`,
                                        args: {
                                            operation: op,
                                            connection_id: this.connectionId,
                                            query_id,
                                            request_id,
                                            socket_id: socketInfo.id,
                                            server_keep_alive_timeout_ms: timeoutMs,
                                        },
                                    });
                                }
                            }
                        }
                    }
                }
                const tryDecompressResponseStream = params.try_decompress_response_stream ?? true;
                const ignoreErrorResponse = params.ignore_error_response ?? false;
                // even if the stream decompression is disabled, we have to decompress it in case of an error
                const isFailedResponse = !(0, client_common_1.isSuccessfulResponse)(_response.statusCode);
                if (tryDecompressResponseStream ||
                    (isFailedResponse && !ignoreErrorResponse)) {
                    const decompressionResult = (0, compression_1.decompressResponse)(_response, log_writer, log_level);
                    if ((0, compression_1.isDecompressionError)(decompressionResult)) {
                        const err = (0, client_common_1.enhanceStackTrace)(decompressionResult.error, currentStackTrace);
                        return reject(err);
                    }
                    responseStream = decompressionResult.response;
                }
                else {
                    responseStream = _response;
                }
                if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
                    log_writer.trace({
                        message: `${op}: response stream created`,
                        args: {
                            operation: op,
                            connection_id: this.connectionId,
                            query_id,
                            request_id,
                            stream_state: {
                                readable: responseStream.readable,
                                readableEnded: responseStream.readableEnded,
                                readableLength: responseStream.readableLength,
                            },
                            is_failed_response: isFailedResponse,
                            will_decompress: tryDecompressResponseStream,
                        },
                    });
                }
                if (isFailedResponse && !ignoreErrorResponse) {
                    try {
                        const errorMessage = await (0, utils_1.getAsText)(responseStream);
                        const err = (0, client_common_1.enhanceStackTrace)((0, client_common_1.parseError)(errorMessage), currentStackTrace);
                        reject(err);
                    }
                    catch (e) {
                        // If the ClickHouse response is malformed
                        const err = (0, client_common_1.enhanceStackTrace)(e, currentStackTrace);
                        reject(err);
                    }
                }
                else {
                    return resolve({
                        stream: responseStream,
                        summary: params.parse_summary
                            ? this.parseSummary(op, _response)
                            : undefined,
                        response_headers: { ..._response.headers },
                        http_status_code: _response.statusCode ?? undefined,
                    });
                }
            };
            function onAbort() {
                // Prefer 'abort' event since it always triggered unlike 'error' and 'close'
                // see the full sequence of events https://nodejs.org/api/http.html#httprequesturl-options-callback
                removeRequestListeners();
                request.once('error', function () {
                    /**
                     * catch "Error: ECONNRESET" error which shouldn't be reported to users.
                     * see the full sequence of events https://nodejs.org/api/http.html#httprequesturl-options-callback
                     * */
                });
                const err = (0, client_common_1.enhanceStackTrace)(new Error('The user aborted a request.'), currentStackTrace);
                reject(err);
            }
            function onClose() {
                // Adapter uses 'close' event to clean up listeners after the successful response.
                // It's necessary in order to handle 'abort' and 'timeout' events while response is streamed.
                // It's always the last event, according to https://nodejs.org/docs/latest-v14.x/api/http.html#http_http_request_url_options_callback
                removeRequestListeners();
            }
            function pipeStream() {
                // if request.end() was called due to no data to send
                if (request.writableEnded) {
                    return;
                }
                const bodyStream = (0, utils_1.isStream)(params.body)
                    ? params.body
                    : stream_1.default.Readable.from([params.body]);
                const callback = (e) => {
                    if (e) {
                        removeRequestListeners();
                        const err = (0, client_common_1.enhanceStackTrace)(e, currentStackTrace);
                        reject(err);
                    }
                };
                if (params.enable_request_compression) {
                    stream_1.default.pipeline(bodyStream, zlib_1.default.createGzip(), request, callback);
                }
                else {
                    stream_1.default.pipeline(bodyStream, request, callback);
                }
            }
            const onSocket = (socket) => {
                try {
                    if (this.params.keep_alive.enabled &&
                        this.params.keep_alive.idle_socket_ttl > 0) {
                        const socketInfo = this.knownSockets.get(socket);
                        // It is the first time we've encountered this socket,
                        // so it doesn't have the idle timeout handler attached to it
                        if (socketInfo === undefined) {
                            const socket_id = this.getNewSocketId();
                            if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
                                log_writer.trace({
                                    message: `${op}: using a fresh socket, setting up a new 'free' listener`,
                                    args: {
                                        operation: op,
                                        connection_id: this.connectionId,
                                        query_id,
                                        request_id,
                                        socket_id,
                                    },
                                });
                            }
                            const newSocketInfo = {
                                id: socket_id,
                                idle_timeout_handle: undefined,
                                usage_count: 1,
                            };
                            this.knownSockets.set(socket, newSocketInfo);
                            // When the request is complete and the socket is released,
                            // make sure that the socket is removed after `idle_socket_ttl`.
                            socket.on('free', () => {
                                if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
                                    log_writer.trace({
                                        message: `${op}: socket was released`,
                                        args: {
                                            operation: op,
                                            connection_id: this.connectionId,
                                            query_id,
                                            request_id,
                                            socket_id,
                                        },
                                    });
                                }
                                // Avoiding the built-in socket.timeout() method usage here,
                                // as we don't want to clash with the actual request timeout.
                                const idleTimeoutHandle = setTimeout(() => {
                                    if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
                                        log_writer.trace({
                                            message: `${op}: removing idle socket`,
                                            args: {
                                                operation: op,
                                                connection_id: this.connectionId,
                                                query_id,
                                                request_id,
                                                socket_id,
                                                idle_socket_ttl_ms: this.params.keep_alive.idle_socket_ttl,
                                            },
                                        });
                                    }
                                    this.knownSockets.delete(socket);
                                    socket.destroy();
                                }, this.params.keep_alive.idle_socket_ttl).unref();
                                newSocketInfo.idle_timeout_handle = idleTimeoutHandle;
                            });
                            const cleanup = (eventName) => () => {
                                const maybeSocketInfo = this.knownSockets.get(socket);
                                // clean up a possibly dangling idle timeout handle (preventing leaks)
                                if (maybeSocketInfo?.idle_timeout_handle) {
                                    clearTimeout(maybeSocketInfo.idle_timeout_handle);
                                }
                                if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
                                    log_writer.trace({
                                        message: `${op}: received '${eventName}' event, 'free' listener removed`,
                                        args: {
                                            operation: op,
                                            connection_id: this.connectionId,
                                            query_id,
                                            request_id,
                                            socket_id,
                                            event: eventName,
                                        },
                                    });
                                }
                                if (log_level <= client_common_1.ClickHouseLogLevel.WARN) {
                                    if (responseStream && !responseStream.readableEnded) {
                                        log_writer.warn({
                                            message: `${op}: socket was closed or ended before the response was fully read. ` +
                                                'This can potentially result in an uncaught ECONNRESET error! ' +
                                                'Consider fully consuming, draining, or destroying the response stream.',
                                            args: {
                                                operation: op,
                                                connection_id: this.connectionId,
                                                query_id,
                                                request_id,
                                                socket_id,
                                                event: eventName,
                                            },
                                        });
                                    }
                                }
                            };
                            socket.once('end', cleanup('end'));
                            socket.once('close', cleanup('close'));
                        }
                        else {
                            clearTimeout(socketInfo.idle_timeout_handle);
                            socketInfo.idle_timeout_handle = undefined;
                            if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
                                log_writer.trace({
                                    message: `${op}: reusing socket`,
                                    args: {
                                        operation: op,
                                        connection_id: this.connectionId,
                                        query_id,
                                        request_id,
                                        socket_id: socketInfo.id,
                                        usage_count: socketInfo.usage_count,
                                    },
                                });
                            }
                            socketInfo.usage_count++;
                        }
                    }
                }
                catch (e) {
                    if (log_level <= client_common_1.ClickHouseLogLevel.ERROR) {
                        log_writer.error({
                            message: `${op}: an error occurred while housekeeping the idle sockets`,
                            err: e,
                            args: {
                                operation: op,
                                connection_id: this.connectionId,
                                query_id,
                                request_id,
                            },
                        });
                    }
                }
                // Socket is "prepared" with idle handlers, continue with our request
                pipeStream();
                // This is for request timeout only. Surprisingly, it is not always enough to set in the HTTP request.
                // The socket won't be destroyed, and it will be returned to the pool.
                if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
                    const socketInfo = this.knownSockets.get(socket);
                    if (socketInfo) {
                        log_writer.trace({
                            message: `${op}: setting up request timeout`,
                            args: {
                                operation: op,
                                connection_id: this.connectionId,
                                query_id,
                                request_id,
                                socket_id: socketInfo.id,
                                timeout_ms: requestTimeout,
                            },
                        });
                    }
                    else {
                        log_writer.trace({
                            message: `${op}: setting up request timeout on a socket`,
                            args: {
                                operation: op,
                                connection_id: this.connectionId,
                                query_id,
                                request_id,
                                timeout_ms: requestTimeout,
                            },
                        });
                    }
                }
                socket.setTimeout(this.params.request_timeout, onTimeout);
            };
            const onTimeout = () => {
                removeRequestListeners();
                if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
                    const socket = request.socket;
                    const maybeSocketInfo = socket
                        ? this.knownSockets.get(socket)
                        : undefined;
                    const socketState = request.socket
                        ? {
                            connecting: request.socket.connecting,
                            pending: request.socket.pending,
                            destroyed: request.socket.destroyed,
                            readyState: request.socket.readyState,
                        }
                        : undefined;
                    const responseStreamState = responseStream
                        ? {
                            readable: responseStream.readable,
                            readableEnded: responseStream.readableEnded,
                            readableLength: responseStream.readableLength,
                        }
                        : undefined;
                    log_writer.trace({
                        message: `${op}: timeout occurred`,
                        args: {
                            operation: op,
                            connection_id: this.connectionId,
                            query_id,
                            request_id,
                            socket_id: maybeSocketInfo?.id,
                            timeout_ms: requestTimeout,
                            socket_state: socketState,
                            response_stream_state: responseStreamState,
                            has_response_stream: responseStream !== undefined,
                        },
                    });
                }
                const err = (0, client_common_1.enhanceStackTrace)(new Error('Timeout error.'), currentStackTrace);
                try {
                    request.destroy();
                }
                catch (e) {
                    if (log_level <= client_common_1.ClickHouseLogLevel.ERROR) {
                        log_writer.error({
                            message: `${op}: An error occurred while destroying the request`,
                            err: e,
                            args: {
                                operation: op,
                                connection_id: this.connectionId,
                                query_id,
                                request_id,
                            },
                        });
                    }
                }
                reject(err);
            };
            function removeRequestListeners() {
                if (request.socket) {
                    request.socket.setTimeout(0); // reset previously set timeout
                    request.socket.removeListener('timeout', onTimeout);
                }
                request.removeListener('socket', onSocket);
                request.removeListener('response', onResponse);
                request.removeListener('error', onError);
                request.removeListener('close', onClose);
                if (params.abort_signal) {
                    request.removeListener('abort', onAbort);
                }
            }
            request.on('socket', onSocket);
            request.on('response', onResponse);
            request.on('error', onError);
            request.on('close', onClose);
            if (params.abort_signal) {
                params.abort_signal.addEventListener('abort', onAbort, {
                    once: true,
                });
            }
            if (!params.body) {
                try {
                    return request.end();
                }
                catch (e) {
                    if (log_level <= client_common_1.ClickHouseLogLevel.ERROR) {
                        log_writer.error({
                            message: `${op}: an error occurred while ending the request without body`,
                            err: e,
                            args: {
                                operation: op,
                                connection_id: this.connectionId,
                                query_id,
                                request_id,
                            },
                        });
                    }
                }
            }
        });
    }
}
exports.NodeBaseConnection = NodeBaseConnection;
const PingQuery = `SELECT 'ping'`;
